plothist/plotters.py

```diff
@@ -291,18 +291,121 @@ def plot_2d_hist(hist, ax, pcolormesh_kwargs={}, colorbar_kwargs={}):
+    ax.get_figure().colorbar(im, cax=ax_colorbar, **colorbar_kwargs)
+
+
+def plot_2d_hist_with_projections(
+    hist,
+    xlabel=None,
+    ylabel=None,
+    ylabel_x_projection=None,
+    colorbar_label=None,
+    save_as=None,
+    pcolormesh_kwargs={},
+    colorbar_kwargs={},
+):
+    """Plot a 2D histogram with projections on the x and y axes.
+
+    Parameters
+    ----------
+    hist : 2D boost_histogram.Histogram
+        The 2D histogram to plot.
+    xlabel : str, optional
+        Label for the x axis. Default is None.
+    ylabel : str, optional
+        Label for the y axis. Default is None.
+    ylabel_x_projection : str, optional
+        Label for the y label of the x projection. Default is None.
+    colorbar_label : str, optional
+        Label for the colorbar. Default is None.
+    save_as : str, optional
+        Path to save the figure to. Default is None.
+    pcolormesh_kwargs : dict, optional
+        Keyword arguments for the pcolormesh call. Default is {}.
+    colorbar_kwargs : dict, optional
+        Keyword arguments for the colorbar call. Default is {}.
+
+    Returns
+    -------
+    fig : matplotlib.figure.Figure
+        The figure.
+    ax_2d : matplotlib.axes.Axes
+        The axes for the 2D histogram.
+    ax_x_projection : matplotlib.axes.Axes
+        The axes for the x projection.
+    ax_y_projection : matplotlib.axes.Axes
+        The axes for the y projection.
+    ax_colorbar : matplotlib.axes.Axes
+        The axes for the colorbar.
+    """
+
+    colorbar_kwargs.setdefault("label", colorbar_label)
+
+    fig, axes = plt.subplots(
+        figsize=(5.25, 5),
+        nrows=2,
+        ncols=3,
+        gridspec_kw={"width_ratios": [4, 1, 0.25], "height_ratios": [1, 4]},
+    )
+
+    ax_2d = axes[1][0]
+    ax_x_projection = axes[0][0]
+    ax_y_projection = axes[1][1]
+    ax_colorbar = axes[1][2]
+    axes[0, 1].axis("off")
+    axes[0, 2].axis("off")
+
+    plot_2d_hist(
+        hist,
+        ax=ax_2d,
+        ax_colorbar=ax_colorbar,
+        pcolormesh_kwargs=pcolormesh_kwargs,
+        colorbar_kwargs=colorbar_kwargs,
+    )
+    plot_hist(hist[:, :: bh.sum], ax=ax_x_projection)
+    plot_hist(hist[:: bh.sum, :], ax=ax_y_projection, orientation="horizontal")
+
+    _ = ax_x_projection.xaxis.set_ticklabels([])
+    _ = ax_y_projection.yaxis.set_ticklabels([])
+
+    ax_y_projection.xaxis.set_ticks_position("top")
+    ax_y_projection.xaxis.set_label_position("top")
+
+    ax_colorbar.yaxis.set_offset_position("left")
+
+    xlim = (hist.axes[0].edges[0], hist.axes[0].edges[-1])
+    ylim = (hist.axes[1].edges[0], hist.axes[1].edges[-1])
+    ax_2d.set_xlim(xlim)
+    ax_x_projection.set_xlim(xlim)
+    ax_2d.set_ylim(ylim)
+    ax_y_projection.set_ylim(ylim)
+
+    ax_2d.set_xlabel(xlabel)
+    ax_2d.set_ylabel(ylabel)
+    ax_x_projection.set_ylabel(ylabel_x_projection)
+
+    hspace = 0.15
+    wspace = 0.15
+    fig.subplots_adjust(hspace=hspace, wspace=wspace)
```

Comment:
The space between the (`ax_2d` and `ax_x_projection`) and (`ax_2d` and `ax_y_projection`) is not the same.

Reply:
Took 3 hours to fix, but I found a way. The cost of this is that the y labels cannot be aligned.